### ISSUE TITLE
string(2) may be too short to identify a language

### ISSUE THREAD

**Issue by lantw44** [ocpi/ocpi#328]
Currently OCPI limits `language` field to ISO 639-1 code. While it may make the specification easier to implement, it means languages which cannot be solely identified by ISO 639-1 codes cannot be properly supported by the protocol.

- A language can be written in multiple scripts. For example, Mandarin Chinese can be written with Traditional Chinese characters and Simplified Chinese characters, but they cannot be distinguished ISO 639-1 code `zh`.
- Multiple languages can share a single ISO 639-1 code. For example, there are separate ISO 639-3 codes assigned to Mandarin Chinese `cmn`, Min Nan Chinese `nan`, Hakka Chinese `hak`, but all of them are `zh` in ISO 639-1 code.
- There are languages which don't have any ISO 639-1 code. For example, none of Formosan languages have ISO 639-1 codes, but they have ISO 639-3 codes.

While the issue probably doesn't affect most languages used on the Internet, I think a protocol intended to be used by companies around the world should not make it harder to support languages with greater varieties or minority languages.

It seems the best way to describe the language of the text is the [language tag](https://tools.ietf.org/html/rfc5646), which can address the problem described above by allowing users to use ISO 639-3 codes and include additional information, such as script, region, variant, in tags. If it turns out to be too complex, the [format used in glibc](https://lh.2xlibre.net/locales/), which  supports ISO 639-3 codes and country codes, may be a simpler alternative.

**Comment by RobertdeLeeuw-IHomer** [comment]
Good point. Something we need to keep in mind for a next version of OCPI.
thank you for pointing this out.

**Comment by lantw44** [comment]
It seems that `language` fields still use `CiAsciiString[2]` and require ISO 639-1 in the review version of OCPI 3.0. Is there any progress on this issue?

**Comment by reinierl** [comment]
No progress.T his was forgotten about. But good point, I'll create an issue for it on the OCPI 3.0 Github repository.

**Comment by reinierl** [comment]
There is no progress.

I think this is because there is currently nobody running into a problem because of this. I don't think any company is trying to share content via OCPI in languages outside 639-2 like Saterland Frisian or such. We tackle issues in order of priority so I don't see this getting adressed until someone actually runs into this limitation.

**Comment by reinierl** [comment]
Oh, I was a bit more pessimistic than I need to be. It's slated for inclusion in 3.0, we've just been distracted by OCPI 2.3.0 development. 

Something like the glibc locale system seems nice because it easily maps to commonsense classifications, it's already used by a lot of other software, and leaves content producers free to use easily understood general codes (like "sr" for Serbian) or very specific ones ("sr_RS@Latin" for Serbian of Serbia written in the Latin script) as they see fit. I don't know how well the language tags system preserves those properties. If it does it seems like a good choice.

**Comment by lantw44** [comment]
> I think this is because there is currently nobody running into a problem because of this.

We did receive data like this from a company in Taiwan:

```json
{
  "language": "zh-TW",
  "text": "本站點充電樁為測試用途，請勿使用"
}
```

This is clearly not a `string(2)`. Unless we want to create and apply a special schema for them, their data can't pass validation. We are forced to truncate their `language` fields before they reach the OCPI layer.

> I don't know how well the language tags system preserves those properties. If it does it seems like a good choice.

Language tags have script and region subtags, so both `RS` and `latin` parts in the example are supported.

**Comment by reinierl** [comment]
Today the issue was discussed in the OCPI Development Working Group and we're indeed going to use language tags in OCPI 3.0.

Thanks for the patience after almost four years 😬

**Comment by lantw44** [comment]
Thank you. We are looking forward to the updated version!
